servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ScanWithMapper.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * Provides the ability to transform (aka map) signals emitted via the {@link Publisher#scanWith(Supplier)} operator.
+ * @param <T> Type of items emitted by the {@link Publisher} this operator is applied.
+ * @param <R> Type of items emitted by this operator.
+ */
+public interface ScanWithMapper<T, R> {
+    /**
+     * Invoked on each {@link Subscriber#onNext(Object)} signal and maps from type {@link T} to type {@link R}.
+     * @param next The next element emitted from {@link Subscriber#onNext(Object)}.
+     * @return The result of mapping {@code next}.
+     */
+    @Nullable
+    R mapOnNext(@Nullable T next);
```

Comment:
> After shifting to JDK functional interfaces I realized this won't work for the intended use case. trailers manipulation needs to keep the state internal and doesn't expose the state externally into the stream. I reverted back to the independent interface approach and clarified naming and javadocs as suggested.

I'm good with a custom interface that covers all use-cases, instead of composing multiple functions. Just debating with myself if this mapper have to be always stateful or can give users more flexibility to decide how to use it.
```java
public interface ScanWithMapper<T, R> {
    @Nullable
    R initialSeed();

    @Nullable
    R mapOnNext(@Nullable R prev, @Nullable T next);

    @Nullable
    R mapOnError(@Nullable R prev, Throwable cause) throws Throwable;

    @Nullable
    R mapOnComplete(@Nullable R prev);
    
    boolean mapTerminal();
}
```
This way users have an option to implement a stateless `ScanWithMapper`, but can also do it in a statefull way if necessary. If the impl tracks its own state which is more complex than just `prev`, they can ignore the first argument. WDYT?

Reply:
I prefer to keep this API simpler and targeted for the use cases at hand (which don't expose internal state in the stream or interface). We can always add an override which internalizes the state (see below), and/or a specialized "stateless" API override so users can choose which specific case they need (as opposed to one API that combines multiple use cases).

```java
    public final <R> Publisher<R> scanWith(Supplier<R> initial,
                                           BiFunction<R, ? super T, R> accumulator,
                                           Predicate<R> mapTerminal,
                                           UnaryOperator<R> onCompleteMapper,
                                           BiFunction<R, Throwable, R> onErrorMapper) {
        return scanWith(() -> new ScanWithMapper<T, R>() {
            @Nullable
            private R state = initial.get();

            @Nullable
            @Override
            public R mapOnNext(@Nullable final T next) {
                state = accumulator.apply(state, next);
                return state;
            }

            @Nullable
            @Override
            public R mapOnError(final Throwable cause) throws Throwable {
                return onErrorMapper.apply(state, cause);
            }

            @Nullable
            @Override
            public R mapOnComplete() {
                return onCompleteMapper.apply(state);
            }

            @Override
            public boolean mapTerminal() {
                return mapTerminal.test(state);
            }
        });
    }
```